feat(android): Add docs for new ANR implementation (#6973)
getsentry/sentry-docs: src/platforms/android/configuration/app-not-respond.mdx

description: "Learn how to turn off or specify ANR."
---

The Android SDK uses two implementations to detect ANRs:

- Reading [ApplicationExitInfo](https://developer.android.com/reference/android/app/ApplicationExitInfo) for devices running Android 11 and above (v2)
- A Watchdog thread for devices running Android version below 11 (v1)

> The new implementation (v2) is more accurate and enables us to be on-par with Google Play Console with regards to reported ANR events and [ANR rate](/platforms/android/performance/instrumentation/performance-metrics/#application-not-responding-anr-rate). We also can capture a thread dump with some additional information like held locks to help you resolve ANRs more efficiently.

Reviewer: `m`: 
It may make sense not to mention it's more accurate but only mention that it's using the same data source as Google Play Console. As you point out, the downside is that you get less context with v2.

Author: sounds good, I'll change the wording

> <Note>
> 
> The v1 approach (Watchdog) still has some advantages over v2, for example, capturing screenshots and transactions with profiles at the time of ANR. Therefore, we're considering support both approaches working alongside each other. Please, upvote this [discussion](https://github.com/getsentry/sentry-java/discussions/2716) and share your feedback, if you have a case for v1 and v2 working together.
> 
> </Note>

Reviewer 2: ```suggestion
The original Watchdog approach (v1) has some advantages over v2, like capturing screenshots and transactions with profiles at the time of ANR.

<Note>

We're considering SDK support for both approaches working alongside each other on Android 11 and up. Please upvote [this GitHub discussion](https://github.com/getsentry/sentry-java/discussions/2716) and share your feedback if you have a case for v1 and v2 working together.

</Note>
```

Reviewer 2: What do you think about splitting only this into the note? I feel like it makes more sense this way.

> 

Reviewer 2: ```suggestion

Both ANR detection implementations are controlled by the same flag:

```xml {filename:AndroidManifest.xml}
<application>
    <meta-data android:name="io.sentry.anr.enable" android:value="false" />
</application>
```

```

    <meta-data android:name="io.sentry.anr.timeout-interval-millis" android:value="2000" />
</application>
```

## ApplicationExitInfo (v2)

> The new approach reads the [ApplicationExitInfo](https://developer.android.com/reference/android/app/ApplicationExitInfo) API on the next app launch
> and asynchronously sends ANR events for each ANR in the history, enriching only the latest one with breadcrumbs, contexts, tags, etc.

Reviewer 2: ```suggestion
This approach reads the [ApplicationExitInfo](https://developer.android.com/reference/android/app/ApplicationExitInfo) API on the next app launch
and asynchronously sends ANR events for each ANR in the history, enriching only the latest one with breadcrumbs, contexts, tags, etc.
```

![ANR](app-not-respond.png)

> The new ANR detection is controlled by the same flag as the old one:
> 
> ```xml {filename:AndroidManifest.xml}
> <application>
>     <meta-data android:name="io.sentry.anr.enable" android:value="false" />
> </application>
> ```

Reviewer 2: ```suggestion
```

Reviewer 2: move this up in the doc
